rfcs/0000-rfc-forms.md

```diff
@@ -0,0 +1,148 @@
+---
+Status: Active
+Champions: christian24
+PR: {{ update_with_pr_number }}
+---
+
+# @lit-labs/forms
+
+Add advanced form validation for web components. This is a followup to the discussion in the lit repository: https://github.com/lit/lit/discussions/2489 
+
+## Objective
+
+Provide an easy to use dev experience for validating forms with web components in them. 
+
+### Goals
```

Comment:
Regarding form-associated elements, I don't think anything is needed, you can use them like any other control. If their APIs are different from a native `<input>` (let's say a `<counter>` has a property named "count" instead of "value"), in my prototype I created a simple `ValueAccessor` which is basically a bridge between the library and the component's properties/events. It doesn't matter if it's a form-associated element, it could be any element with any properties or events, once you have the bridge.

Here's what they do and why I don't think this solution would interfere with them:

- When you place a form-associated custom element inside a <form>, it's automatically associated with the form, like a browser-provided control. 
 
**[In this library we wouldn't use the browser's form instance so it doesn't matter]**

- The element can be labeled using a <label> element.

**[No problem here]**

- The element can set a value that's automatically submitted with the form.

**[Form submission is part of the "progressive enhancement" side that's fundamentally divergent from this solution, in other words we always have a model with all the values and we submit what we want via AJAX]**

- The element can set a flag indicating whether or not it has valid input. If one of the form controls has invalid input, the form can't be submitted.

**[same as above]**

- The element can provide callbacks for various parts of the form lifecycle—such as when the form is disabled or reset to its default state.

**[I think there's no problem here]**

- The element supports the standard CSS pseudoclasses for form controls, like :disabled and :invalid.

**[No problem here]**

Reply:
Agree with @UserGalileo in general I think the browser validation API just doesn't go far enough, so this is completely separate from any browser API.